### ISSUE TITLE
Xamarin Android crash

### ISSUE THREAD

**Issue by mattheworiordan** [ably/ably-dotnet#96]
A customer reported this issue on Xamarin:

```
System.NullReferenceExceptionObject reference not set to an instance of an object
Raw
SuperSocket.ClientEngine.ConnectAsyncExtension.SocketConnectCompleted(object sender, SocketAsyncEventArgs e)<2b3384da5fba458d95aaf3e5bd2b26a7>:0
SuperSocket.ClientEngine.ConnectAsyncExtension.OnGetHostAddresses(IAsyncResult result)<2b3384da5fba458d95aaf3e5bd2b26a7>:0
System.Runtime.Remoting.Messaging.AsyncResult.Invoke(AsyncResult)(wrapper managed-to-native)
System.Runtime.Remoting.Messaging.AsyncResult.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()<3dc9ed3d31194319991e686734adcb10>:0
System.Threading.ThreadPoolWorkQueue.Dispatch()<3dc9ed3d31194319991e686734adcb10>:0
System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()<3dc9ed3d31194319991e686734adcb10>:0
```

**Comment by mattheworiordan** [comment]
Internally this maps to ticket 1582

**Comment by marto83** [comment]
@mattheworiordan do you have more info. I found the code for SuperSocket but from the stack trace it's not obvious at all why it failed. 

How often does it happen and is it reproducible at all?

**Comment by mattheworiordan** [comment]
Ok, will ask the customer to add details here. Thanks @marto83

**Comment by NVentimiglia** [comment]
It happens a few times per week. Crashes reports via insights.

**Comment by mattheworiordan** [comment]
Hi @NVentimiglia. Thanks for getting in touch.  Clearly we want to get to the bottom of this ASAp. Are you able to provide any more info at that may help such as what event caused this, any other log entries that may be related etc?

**Comment by NVentimiglia** [comment]
I do not have repo steps at this time, only exceptions caught from our crash manager.

**Comment by NVentimiglia** [comment]
New stack trace from production

```
System.NullReferenceExceptionObject reference not set to an instance of an object
Raw
WebSocket4Net.WebSocket.ClearTimer()
WebSocket4Net.WebSocket.FireClosed()
WebSocket4Net.WebSocket.OnClosed()
WebSocket4Net.WebSocket.client_Error(object sender, ErrorEventArgs e)
SuperSocket.ClientEngine.ClientSession.OnError(Exception e)
SuperSocket.ClientEngine.SslStreamTcpSession.OnDataRead(IAsyncResult result)
System.Net.LazyAsyncResult.Complete(IntPtr userToken)
System.Net.LazyAsyncResult.ProtectedInvokeCallback(object result, IntPtr userToken)
System.Net.LazyAsyncResult.InvokeCallback(object result)
Mono.Net.Security.AsyncProtocolRequest.StartOperation()
Mono.Net.Security.AsyncProtocolRequest.<StartOperation>m__0(object _)
System.Threading.QueueUserWorkItemCallback.WaitCallback_Context(object state)
System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, object state, bool preserveSyncCtx)
System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, object state, bool preserveSyncCtx)
System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem()
System.Threading.ThreadPoolWorkQueue.Dispatch()
ObjCRuntime.Runtime.ThreadPoolDispatcher(Func<TResult> callback)
System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()
```

**Comment by marto83** [comment]
Thanks @NVentimiglia I've created a related issue with the Websocket4Net library. I'll keep you updated when I hear more. I haven't managed to recreated it locally at the moment.

**Comment by marto83** [comment]
@NVentimiglia what version of the ably library do you use?

**Comment by NVentimiglia** [comment]
<package id="ably.io" version="0.8.4" targetFramework="MonoAndroid70" />
<package id="WebSocket4Net" version="0.14.1" targetFramework="MonoAndroid70" />

@marto83

**Comment by NVentimiglia** [comment]
Also picked up a new crash

System.InvalidOperationExceptionAn attempt was made to transition a task to a final state when it had already completed.
Raw
null
System.InvalidOperationExceptionAn attempt was made to transition a task to a final state when it had already completed.
System.Threading.Tasks.TaskCompletionSource<TResult>.SetResult(TResult result)<9413b069249c4ca78f27bc3302b39016>:0
IO.Ably.Transport.TaskWrapper.Callback(bool res, ErrorInfo ei)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.ChannelAwaiter.Fail(ErrorInfo error)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.RealtimeChannel.HandleStateChange(ChannelState state, ErrorInfo error, ProtocolMessage protocolMessage)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.RealtimeChannel.SetChannelState(ChannelState state, ErrorInfo error, ProtocolMessage protocolMessage)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.RealtimeChannel.<OnDetachTimeout>b__57_0()<99060ba189e94f839f7ba4286b72b508>:0
System.Threading.Tasks.Task.InnerInvoke()<9413b069249c4ca78f27bc3302b39016>:0
System.Threading.Tasks.Task.Execute()<9413b069249c4ca78f27bc3302b39016>:0

**Comment by marto83** [comment]
@NVentimiglia Thanks for the crash report. I'll add extra checks to make sure this doesn't happen.

**Comment by NVentimiglia** [comment]
New crash report, any progress ? @marto83 

`System.InvalidOperationExceptionAn attempt was made to transition a task to a final state when it had already completed.
Raw
null
System.InvalidOperationExceptionAn attempt was made to transition a task to a final state when it had already completed.
System.Threading.Tasks.TaskCompletionSource<TResult>.SetResult(TResult result)<f1b1793f200d4ecc9ba0b6ed8e6cdeb8>:0
IO.Ably.Transport.TaskWrapper.Callback(bool res, ErrorInfo ei)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.ChannelAwaiter.Fail(ErrorInfo error)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.RealtimeChannel.HandleStateChange(ChannelState state, ErrorInfo error, ProtocolMessage protocolMessage)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.RealtimeChannel.SetChannelState(ChannelState state, ErrorInfo error, ProtocolMessage protocolMessage)<99060ba189e94f839f7ba4286b72b508>:0
IO.Ably.Realtime.RealtimeChannel.InternalOnInternalStateChanged(object sender, ConnectionStateChange connectionStateChange)<99060ba189e94f839f7ba4286b72b508>:0
System.EventHandler<ConnectionStateChange>.invoke_void_object_TEventArgs(object, ConnectionStateChange)(wrapper delegate-invoke)
IO.Ably.Realtime.Connection.UpdateState(ConnectionStateBase state)<99060ba189e94f839f7ba4286b72b508>:0
at IO.Ably.Transport.ConnectionManager+<>c__DisplayClass51_0+<<SetState>b__0>d.MoveNext () [0x0011d] in <99060ba189e94f839f7ba4286b72b508>:0`

**Comment by marto83** [comment]
Hi @NVentimiglia. I found what is causing the issue and will release a new version of the package today. Thank you for the patience. 

I'll update this issue once that's done.

**Comment by marto83** [comment]
@NVentimiglia I released v0.8.4.2 which will take care of the issue with `attempt was made to transition a task to a final state`.

**Comment by mattheworiordan** [comment]
Thanks @marto83

**Comment by marto83** [comment]
The issue is fixed from 0.8.5 onwards.
